### ISSUE TITLE
RegExp.prototype.exec() and test() parameters are optional

### ISSUE THREAD

**Issue by newlukai** [microsoft/TypeScript#37211]
**TypeScript Version:**
3.8.3

**Search Terms:**
RegExp

**Code**
```ts
import * as assert from 'assert';

let matches = /ab(c)/.exec('abc');
assert.ok(Array.isArray(matches));
assert.equal((matches as RegExpExecArray).length, 2);

matches = /ab(c)/.exec();
assert.equal(matches, null);
```

**Expected behavior:**
Compiles.

**Actual behavior:**
Doesn't compile due to an expected argument for `exec` in line 8. But the corresponding JavaScript code runs without error.

**Playground Link:**
http://www.typescriptlang.org/play/?ssl=8&ssc=29&pln=1&pc=1#code/JYWwDg9gTgLgBAKjgQwM4tagprOAzKCEOAcjW1hIG4BYAKHoBst4RkYBjACy3QF44AemQAjABQcAlIIB0WAB5YOYsiI4lJtOuRwwZEANZiAglCjIAnjOCpT5i2LaceqSZvo7YcgI4BXZIxijuzcvBhwAEpYAOYAovJg8Up2lpIyzAB20TBcADRwAEzuDHROofxCohLScorKxZ56WH4Bwc68+Rm+jIyaQA

**Related Issues:**
Did not find other bugs that looked similar.

**Comment by ilogico** [comment]
It runs without errors but it's certainly a mistake.

**Comment by IllusionMH** [comment]
While it works in JS, JS also has no problems with `null`, `123` and `{a: 123}` (doesn't throw an error), how implicit `undefined` is better?

Just use `.exec('undefined')` for same effect but at least it won be confusing when `/und/.exec() != null` and `/und/.test() === true`.

**Comment by RyanCavanaugh** [comment]
See https://stackoverflow.com/questions/41750390/what-does-all-legal-javascript-is-legal-typescript-mean

**Comment by newlukai** [comment]
I proposed this because I had an issue when using an optional chain:
`/abc/.exec(someObject.someOptionalSubObject?.someValue);`
What TypeScript defines as "JS is improper about this" forces me to write
`/abc/.exec((someObject.someOptionalSubObject?.someValue as string))`.

I'm wondering if that's intended. I could live with it, but I think it's an aspect where TS feels more challenging.

**Comment by ilogico** [comment]
Like @IllusionMH suggested, you might as well write `/abc/.exec(someObject.someSubObject?.someValue ?? 'undefined')`, which is probably not what you meant, but it's equivalent and more explicit.
That code is wrong and TypeScript is warning you about it. You're using casts to eliminate the warning instead of admitting and correcting the mistake :)
You probably mean something like `someObject.someSubObject?.someValue != null ? /abc/.exec(someObject.someSubObject.someValue) : null`

**Comment by newlukai** [comment]
I guess you are right.
